Add findsecbugs plugin to spotbugs.

jeffret-b opened the pull request:
Add findsecbugs plugin to spotbugs. And suppress existing warnings. We should clean some of them up, but that's for different PRs at a later time.

Some of the issues findsecbugs just misidentifies. With some it doesn't understand enough context.

As always with these sort of of analysis tools, adding them to existing code correctly can be a bit of a challenge. It's easier to suppress things that aren't really problems than to clean all of them up. Most of the benefit from applying these tools comes from checking newly added code and preventing new issues from creeping in.

### Proposed changelog entries

* Developer: Add findsecbug plugins to spotbugs build plugin.

### Submitter checklist

- [n/a ] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
The findsecbugs plugin provides a form of testing. The PR changes no production behavior. Existing tests serve to verify consistent behavior.
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

jeffret-b committed:
Add findsecbugs plugin to spotbugs.

And suppress existing warnings. We should clean some of them up, but that's for different PRs.

jvz commented on core/src/main/java/hudson/Proc.java:
```diff
@@ -215,6 +216,7 @@ public LocalProc(String[] cmd,String[] env,InputStream in,OutputStream out, File
          * @param err
          *      null to redirect stderr to stdout.
          */
+        @SuppressFBWarnings("COMMAND_INJECTION")
```
Worth leaving a note that this is the point of this class?

jvz reviewed (approved):
Changes look reasonable.

oleg-nenashev reviewed (dismissed):
please do not suppress warnings without explicit justification.

jeffret-b committed:
Make sure everything has a justification.

And move a couple of items to the exclude file. If we ever get to a lower threshold, these will show up all over the place.

jeffret-b commented:
Added justifications for anything that was missing.

res0nance commented on core/src/main/java/hudson/cli/Connection.java:
```diff
@@ -113,6 +114,7 @@ public void writeObject(Object o) throws IOException {
     /**
      * Receives an object sent by {@link #writeObject(Object)}
      */
+    @SuppressFBWarnings(value = "OBJECT_DESERIALIZATION", justification = "Not used. We should just remove it.")
```
If it Isn't used can we just remove it?

jeffret-b replied:
I'm in favor of that, but there is generally a lot of resistance to removing any public API. The class has been deprecated for almost a year, so maybe we can remove it now. I would put that in a separate PR, though it could be reasonable to put it here.

res0nance commented on core/src/main/java/hudson/cli/Connection.java:
```diff
@@ -192,6 +194,7 @@ public KeyAgreement diffieHellman(boolean side, int keySize) throws IOException,
      * @return
      *      A new {@link Connection} object that includes the transport encryption.
      */
+    @SuppressFBWarnings(value = "STATIC_IV", justification = "Not used. We should just remove it.")
```
Perhaps start deprecating this if we should remove it

jeffret-b replied:
The class has been deprecated for almost a year.

res0nance commented on core/src/main/java/hudson/util/FormFieldValidator.java:
```diff
@@ -145,6 +146,7 @@ public final void process() throws IOException, ServletException {
     /**
      * Gets the parameter as a file.
      */
+    @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "Not used.")
```
Similarly here, if its not used can this be removed?

jeffret-b replied:
The class was deprecated as of 1.294 and tagged with the attribute 5 years ago. I'd be in favor of removing it, though possibly in a different PR.

res0nance commented on core/src/main/java/hudson/util/FormFieldValidator.java:
```diff
@@ -318,6 +320,7 @@ public HudsonURL(StaplerRequest request, StaplerResponse response) {
             super(request, response);
         }
 
+        @SuppressFBWarnings(value = "URLCONNECTION_SSRF_FD", justification = "Not used.")
```
Here as well

res0nance commented on core/src/main/java/hudson/util/FormFieldValidator.java:
```diff
@@ -513,6 +516,7 @@ public Executable(StaplerRequest request, StaplerResponse response) {
             super(request, response, true);
         }
 
+        @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "Not used.")
```
Once more here

madchap commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -78,6 +78,7 @@
 
 import static org.apache.commons.io.FilenameUtils.getBaseName;
 
+@SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "Admistrator action installing a plugin, which could do far worse.")
```
Spell check -- Admi_ni_strator

jeffret-b replied:
```suggestion
@SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "Administrator action installing a plugin, which could do far worse.")
```

jeffret-b committed:
Correct spelling

res0nance reviewed (approved):
👍 Seem good just need the follow up with removals

jeffret-b commented:
I added a PR to remove the deprecated, unused `Connection` class: #4386. We'll see if we can get approvals and traction for these API changes and then we don't need to suppress the findsecbugs findings.

oleg-nenashev commented:
I plan to merge it tomorrow if no negative feedback

jtnord commented on core/src/main/java/hudson/Util.java:
```diff
@@ -595,6 +596,8 @@ public static String ensureEndsWith(@CheckForNull String subject, @CheckForNull
      */
     @Nonnull
+    @SuppressFBWarnings(value = "WEAK_MESSAGE_DIGEST_MD5", justification =
+            "This method should only be used for non-security applications where their weakness is not a problem.")
```
this is wrong.  
Whilst "this method should only be used for non security calls" supressing it here will mask those that do make security calls.

jeffret-b replied:
This really highlights the difficulties of adding something like findsecbugs into existing, legacy systems like Jenkins. What we really should do is to fix everything so that there are no usages of weak message digests like MD5. That's a large project that requires very focused work and consideration of migration or incompatibility. Making the introduction of findsecbugs dependent on that change means in practice that findsecbugs will never be introduced and new code will not be checked. Not introducing findsecbugs makes it less likely that any usages will be checked and any improper usages corrected. Tagging it with the `@SuppressFBWarnings` at least points out that it **shouldn't** be used for security.

Not suppressing it here (i.e., not using findsecbugs) only results in better masking it for anything that already misuses this to make security calls or who might in the future.

We could improve the message somehow. I'm not sure what it would say better.

jtnord replied:
> What we really should do is to fix everything so that there are no usages of weak message digests like MD5

There is nothing wrong with MD5 when used correctly.
using it for a file checksum to check a file is downloaded from a HTTP**S** source and has not been corrupted (because the IP layers will not detect more than a small level of corruption) is perfectly acceptable.  Sure we could go on and use something more secure - but my point is that it is the caller here and this code is correct when used in the previous scenario but NOT when used to check some security related things.  Suppressing this means you will never find those.

The only thing that is safe to do is to remove all helper methods and duplicate code.  this seems bad.

jeffret-b replied:
No, it's not all that bad. It's been common knowledge for a long time that Jenkins uses MD5 and that MD5 has serious flaws. It's suitable for some applications, but not for cryptographic use. It might be okay for fingerprinting but even there we would be better off moving to SHA-2. Our use of MD5 is certainly not good but no one has said it's really bad and needs immediate correction.

It would be really bad if we wanted to support certain high-security environments, such as FIPS. That's not something we're actively pursuing.

Our options now are basically:
1) Close this PR without merging. Nothing changes. We never find the potential problems.
2) Suppress it for now and move forward. Try to fix code and remove suppressions when we can.
3) Replace all of the MD5 usages in Jenkins that need it. (It would be simplest to replace all of them than to analyze which ones might be okay. Besides FIPS requires it.) Introduce findsecbugs only after all of those are fixed. This option is essentially the same as `#1`, because replacing MD5 isn't going to happen while this PR is relevant.
4) Move the suppression into the exclude file.

`#2` is the best option at this point. Suppressing it gives us a (slightly) better chance of getting around to fixing this because it leaves a marker of the issue. Suppressing it leaves a note that other developers _might_ see helping them know not to use this method or to move away from its use.

jtnord replied:
just to be clear I am not blocking this PR.  just pointing out potential issues in that if we go with this with the say Jenkins is written and will be written in the future we will (I beleive) end up with a huge amount of suppressions just to keep the tool happy and I have a doubt over if it will prevent issues in new code.

> Suppress it for now and move forward. Try to fix code and remove suppressions when we can.

you can't easily do this given that it is a public method used by plugins, you can;t just change that contract unless you want to fix all callers in plugins, or we start actually breaking things intentionally in the name of progress (but I am still missing that direction)

jeffret-b replied:
Yes, someday if we want to fix this specific issue, use of MD5, we will probably need to break things. If we want to someday support secure environments like FIPS, we certainly will. According to [Wikipedia](https://en.wikipedia.org/wiki/MD5), "As of 2019, MD5 continues to be widely used, in spite of its well-documented weaknesses and deprecation by security experts." I guess that means we're certainly not alone is continuing to use it at least. Small consolation I suppose.

I definitely find that tools like spotbugs helps produce cleaner code and help identify and prevent potential problems. I've found value from the findings generated by findsecbugs. I just wish it had been in place much earlier to point out some things that could have been improved much easier earlier.

Thanks for the discussion.

jtnord reviewed (commented):
I do not think as is this is a good thing.
the first annotation I found actually hides potentially valid issues by placing the annotation on the wrong method.

if the tool is not good enough to put annotations on the callers rather than the method then the benefit that it offers is negligable.  (the tool did not find any valid issues from what I can see)

jeffret-b commented:
@daniel-beck @Wadeck

Wadeck reviewed (commented):
In order to move forward with the plugin (I also think it seems useful), I propose that we are separating the suppression in two categories, the ones that are real suppressions (like "Client-side code doesn't involve SSRF.") and the ones that we need to care about (like the MD5 and "not used" code).

We need to have all of them to "keep the tool happy" right now, but we should create a follow-up ticket, linked with every suppression that requires changes and to iterate on them. IMO the worse thing to happen is to just add new suppressions without that distinction.

IOW, I propose to add a `TODO JENKINS-XXX remove that method` or equivalant to move forward.

@jtnord how do you feel about this proposal?

jtnord commented:
> @jtnord how do you feel about this proposal?

go for it.  I would also like to see the annotations placed at the correct finest scope where possible and not on class / methods if avoidable.  (that way if code is refactored it means it is harder to end up with a suppression that is actually fixed)

jeffret-b committed:
Correct spelling

Co-Authored-By: Wadeck Follonier <Wadeck@users.noreply.github.com>

jeffret-b committed:
Narrow the suppression.

And add TODO connected with JENKINS tickets for relevant issues.

daniel-beck commented on core/src/main/java/hudson/Util.java:
```diff
@@ -617,6 +620,13 @@ public static String getDigestOf(@Nonnull InputStream source) throws IOException
+    // TODO JENKINS-60563 remove MD5 from all usages in Jenkins
+    @SuppressFBWarnings(value = "WEAK_MESSAGE_DIGEST_MD5", justification =
+            "This method should only be used for non-security applications where the MD5 weakness is not a problem.")
+    private static MessageDigest getMd5() throws NoSuchAlgorithmException {
```
Good time to deprecate this and all callers?

jeffret-b replied:
Deprecating this makes sense. I've wondered about deprecating `Util.getDigestOf()`. Would it make sense to change the implementation to SHA-2? The documentation specifies MD5, but it might make sense to change that along with the implementation. It wouldn't be the first JavaDoc in Jenkins that is out-of-date, wrong, or not helpful.

jeffret-b replied:
I've figured it made more sense to answer that question when engaged in an effort to remove MD5.

jvz replied:
By the time we fully remove MD5, maybe SHA-4 will be a thing by then! We can defer implementation until then as you suggest. :)

daniel-beck commented on core/src/main/java/hudson/security/BasicAuthenticationFilter.java:
```diff
@@ -162,15 +163,20 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
         d.include(req,rsp);
     }
 
+    @SuppressFBWarnings(value = "UNVALIDATED_REDIRECT", justification = "Redirect is validated as processed.")
```
I don't understand this comment.

jeffret-b replied:
I'll see if I can investigate and improve this when I get a chance.

daniel-beck commented on core/src/main/java/jenkins/model/RunIdMigrator.java:
```diff
@@ -314,13 +315,19 @@ public static void main(String... args) throws Exception {
         new RunIdMigrator().unmigrateJobsDir(jobs);
     }
+
+    @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "Only used as an command-line process.")
```
```suggestion
    @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "Only invoked from the command line as a standalone utility")
```

jeffret-b committed:
Address review comments.

jeffret-b commented:
@Wadeck, when you get a chance could you take a look at this again? Thanks!

jeffret-b commented:
@Wadeck could you take a quick look at this?

Wadeck reviewed (approved):
Sorry for the delay, the change is fine, especially with the two tickets ([JENKINS-60562](https://issues.jenkins-ci.org/browse/JENKINS-60562), [JENKINS-60563](https://issues.jenkins-ci.org/browse/JENKINS-60563)).

Thank you @jeffret-b for the consolidation of all those justification / refactoring to have the possibility to include a new quality tooling to our belt!

jeffret-b commented:
With @Wadeck's review we should be in a good state for merging this on in. There isn't any reason to keep the "on-hold" or "needs-more-reviews" labels. We can add the "ready-for-merge" label. Any thoughts @oleg-nenashev or anyone else?

I'll make those label changes tomorrow if there aren't any objections.

timja commented:
> With @Wadeck's review we should be in a good state for merging this on in. There isn't any reason to keep the "on-hold" or "needs-more-reviews" labels. We can add the "ready-for-merge" label. Any thoughts @oleg-nenashev or anyone else?
> 
> I'll make those label changes tomorrow if there aren't any objections.

once the conflict is resolved ^.^

jeffret-b commented:
Good point. I'll fix that in a little bit.

jeffret-b committed:
Remove suppression that were inadvertently merged back in.

jeffret-b commented:
Resolved the merge conflict. If I can get a successful CI run this should be good to go.

timja commented:
I plan to merge this tomorrow if there's no negative feedback

recena commented on pom.xml:
```diff
@@ -407,6 +407,20 @@ THE SOFTWARE.
+              <plugin>
+                <groupId>com.h3xstream.findsecbugs</groupId>
+                <artifactId>findsecbugs-plugin</artifactId>
+                <version>1.10.1</version>
```
**Only** for this special case, I would use `LATEST`

jtnord replied:
`LATEST` is never a good idea as it is not what you think it is...
